### PR TITLE
feat: hosting and disconnect smoke test

### PR DESCRIPTION
**Pull request by fernando-cortez**
### Description (*)
This PR introduces a hosting and disconnecting smoke test. This Playmode test will invoke the same methods as UI elements would normally and progress the player through the normal game flow. Along the steps, creation of a host will be validated, as well as a networked scene load, and disconnection.

Custom test parameters allow this test to be replicated for all character classes.

### Issue Number(s) (*)

Fixes issue(s): [MTT-2414](https://jira.unity3d.com/browse/MTT-2414)
### Manual testing scenarios

1. Open up the TestRunner window, go into Playmode tests tab and run the test.
### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by SamuelBellomo** [comment]
I'm guessing this will need some rewrite with lobby being merged correct?

File: Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+            {
+                timer -= Time.deltaTime;
+                yield return null;
+            }
+
+            clientMainMenuState.OnStartClicked();
+
+            yield return new WaitForEndOfFrame();
+
+            lobbyUIMediator.ToggleCreateLobbyUI();
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:73]: I'm noticing you're doing a lot of these manual steps. Shouldn't clicking the Start button do this for you?

Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:73]: If you're doing these manual steps, then why not just trigger hosting using GameNetPortal and go from there?

File: Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+            yield return GoToBossRoomScene();
+
+            yield return Disconnect();
+        }
+
+        IEnumerator GoToMainMenuScene()
+        {
+            // load Bootstrap scene
+            SceneManager.LoadSceneAsync(k_BootstrapSceneName);
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:124]: I'm a bit confused at what we're trying to test here.
The actual in game flow would use LP's scene wrapper, why the direct call here?
If we're doing unit tests, we should call lower level methods individually, with no dependencies and make sure to only test that unit.
If we're doing integration tests, we should use as little custom code to move things around and rely on in-game flow code already present in game.

File: Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            uiSettingsCanvas.OnClickQuitButton();
+
+            yield return new WaitForFixedUpdate();
+
+            var uiQuitPanel = GameObject.FindObjectOfType<UIQuitPanel>(true);
+            Assert.That(uiQuitPanel != null, $"{nameof(UIQuitPanel)} component not found!");
+            uiQuitPanel.Quit();
+
+            // TODO: validate with SDK why this is still needed
+            yield return new WaitForSeconds(1f);
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:188]: shouldn't we just wait one frame? yield return null?

File: Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+                yield return null;
+            }
+
+            Assert.IsTrue(AuthenticationService.Instance.IsAuthorized);
+
+            clientMainMenuState.OnDirectIPClicked();
+
+            yield return new WaitForEndOfFrame();
+
+            ipUIMediator.ToggleCreateIPUI();
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:234]: same here

File: Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+            Assert.That(uiSettingsCanvas != null, $"{nameof(UISettingsCanvas)} component not found!");
+            uiSettingsCanvas.OnClickQuitButton();
+
+            yield return new WaitForFixedUpdate();
+
+            var uiQuitPanel = GameObject.FindObjectOfType<UIQuitPanel>(true);
+            Assert.That(uiQuitPanel != null, $"{nameof(UIQuitPanel)} component not found!");
+            uiQuitPanel.Quit();
+
+            // TODO: validate with SDK why this is still needed
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:187]: as discussed, MTT ticket in comments please

File: Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs (hunk 1 above)
Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:73]: as discussed, some comments on what these do would be great :)

File: Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            yield return null;
+
+            // verify hosting is successful
+            Assert.That(m_NetworkManager.IsListening);
+
+            // CharSelect is loaded as soon as hosting is successful, validate it is loaded
+            yield return GoToCharacterSelection(playerIndex);
+
+            // selecting ready as host with no other party members will load BossRoom scene; validate it is loaded
+            yield return GoToBossRoomScene();
```
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Tests/Runtime/HostAndDisconnectTest.cs:117]: as discussed, these "GoTo" something should be "wait for go to"

**Review by LPLafontaineB — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Review by LPLafontaineB — APPROVED** [review]

**Review by LPLafontaineB — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Pull request merged**